doc/reference.rst

```diff
@@ -323,3 +323,193 @@ C++
+  from ``Print::``) function is also usually blocking until the full buffer is
+  transmitted. Both functions return the number of transmitted bytes.
+
+  That's the way the Stream class works and is commonly used.
+
+  Classes derived from ``Stream::`` also usually introduce the ``read(buffer,
+  len)`` method, which is similar to ``readBytes(buffer, len)`` without
+  timeout: the returned value can be less than the requested size, so special
+  care must be taken with this function, introduced in the Arduino
+  ``Client::`` class.  This function has also been introduced in other classes
+  not derivating from ``Client::``, e.g.  ``HardwareSerial::``.
+
+Stream extensions
+
+  Proposed Stream extensions are designed to be compatible with Arduino API,
+  and offer a additional methods to make transfers more efficients and easier
+  to use.
+
+  The serial to network transfer above can be written like this:
+
+  .. code:: cpp
+
+    serverClient.toNow(Serial); // chunk by chunk
+    Serial.toNow(serverClient); // chunk by chunk
+
+  An echo service can be written like this:
+
+  .. code:: cpp
+
+    serverClient.toNow(serverClient); // tcp echo service
+
+    Serial.toNow(Serial);             // serial software loopback
+
+  Beside reducing coding time, these methods optimize transfers by avoiding
+  buffer copies when possible.
+
+  - User facing API: ``Stream::to()``
+
+    The goal of streams is to transfer data between producers and consumers,
+    like the telnet/serial example above.  Four methods are proposed, all of
+    them return the number of transmitted bytes:
+
+    - ``Stream::toSize(dest, size [, timeout])``
+
+      This method waits up to the given or default timeout to transfer
+      ``size`` bytes to the the ``dest`` Stream.
+
+    - ``Stream::toUntil(dest, delim [, timeout])``
+
+      This method waits up to the given or default timeout to transfer data
+      until the character ``delim`` is met.  The delimiter is read but not
+      transfered.
+
+    - ``Stream::toNow(dest)``
+
+      This method transfers all already available data to the destination.
+      There is no timeout and the returned value is 0 when there is nothing
+      to transfer or no room in the destination.
+
+    - ``Stream::toAll(dest [, timeout])``
+
+      This method waits up to the given or default timeout to transfer all
+      available data.  It is useful when source is able to tell that no more
+      data will be available for this call, or when destination is able to
+      tell that it will no more be able to receive.
+
+      For example, a source String will not grow during the transfer, or a
+      particular network connection supposed to send a fixed amount of data
+      before closing.  ``::toAll()`` will receive all bytes.  Timeout is
+      useful when destination needs processing time (e.g.  network or serial
+      input buffer full = please wait a bit).
+
```

Comment:
This is another style/naming thing, so take it with a grain of salt, but the new method names are puzzling to me.

Instead of `Stream::toXXX` I would suggest `Stream::sendXXX` which is really what's going on (i.e. the verb, not the preopsition).
````
Stream::sendCount(dest, count); // Send count bytes to dest
Stream::sendUntilChar(dest, chr); // Send data to dest until chr matched
Stream::sendAvailable(dest); // Send all presently available data to dest
Stream::sendUntilEnd(); // Send all data to dest, even if source reads block
````